File: frontend/src/api/mail_followup.api.ts
import { http } from "./http";
import type { ContactField } from "./call_list.api";

/**
 * Der Mailversand – was aus einer Zusage der Telefonakquise geworden ist.
 *
 * Die Zeilen sind die Kontakte der Telefonakquise im Zustand „zugesagt"; neu
 * ist allein der Versandzustand. Die Seite hat deshalb eine eigene
 * Berechtigung, aber keine eigenen Kontakte: wird eine Zusage drüben
 * richtiggestellt, verschwindet die Zeile hier von selbst.
 */

/**
 * Die Zustände einer Zusage im Versand. Spiegel von `MailState` im Backend
 * (`backend/app/schemas/mail_followup.py`).
 *
 * Beschriftung, Beschreibung und Tonlage der Knöpfe kommen dagegen *mit der
 * Antwort* (`MailBoard.actions`) – hier stehen nur die IDs, weil das Frontend
 * jeder ID ein Symbol zuordnet. `mailStates.test.ts` hält beide Seiten
 * zusammen, nach dem Muster von `callOutcomes.test.ts`.
 */
export const MAIL_STATES = [
  "offen",
  "versendet",
  "nachfassen",
  "nachgefasst",
  "positiv",
  "abgelehnt",
  "keine_antwort",
] as const;

export type MailState = (typeof MAIL_STATES)[number];

/**
 * Die Bau-Einschätzung einer Zusage. Spiegel von `BuildReadiness` im Backend.
 *
 * Zweite, vom Versandstand unabhängige Größe: sie sagt, wo die *Website*
 * steht. Zuerst, ob die bestehende Inhalt hat, den die neue übernehmen kann –
 * hat sie keinen, ist es kein Redesign mehr, dann müssen Texte entstehen, und
 * darüber muss vorher jemand mit dem Kunden sprechen. Und dann, mit
 * `in_development` und `ready_to_mail`, dass die neue Seite gebaut wird bzw.
 * schon steht und nur noch zum Betrieb muss: gebaut, aber noch nicht
 * verschickt.
 *
 * `unbewertet` ist der Ausgangswert *und* der Rückweg: entfernt wird ein
 * Marker, indem man ihn setzt (siehe `MailUpdate.readiness`).
 */
export const BUILD_READINESS = [
  "unbewertet",
  "ready_to_build",
  "in_development",
  "ready_to_mail",
  "missing_content",
] as const;

export type BuildReadiness = (typeof BUILD_READINESS)[number];

/**
 * Ein Marker, wie ihn das Frontend rendert – Beschriftung und Beschreibung
 * kommen mit der Antwort (`MailBoard.readiness_options`).
 *
 * Ohne `tone`, anders als bei den Zustands-Knöpfen: „Missing Content" ist
 * keine schlechte Nachricht, sondern mehr Arbeit. Die Farben der Werte stehen
 * deshalb in der Oberfläche (`READINESS`).
 */
export interface ReadinessOptionInfo {
  id: BuildReadiness;
  label: string;
  description: string;
}

/**
 * Der Umfangs-Marker („Bigger than expected"), wie ihn das Backend mitschickt.
 *
 * Dritte Größe neben Versandstand und Bau-Einschätzung – und die einzige, die
 * mit beiden *kombinierbar* ist: eine Seite kann „In Development" und größer
 * als ein Onepager sein, und genau dafür gibt es sie. Deshalb ein eigenes
 * Feld und kein weiterer `BuildReadiness`-Wert, der beim nächsten Bauschritt
 * wieder verschwände.
 *
 * Ein Eintrag statt eines Katalogs: der Marker ist gesetzt oder nicht.
 * `undo_description` ist der Titel des gesetzten Knopfes – dieselbe Rolle wie
 * „Einschätzung entfernen" im Katalog der Bau-Einschätzung.
 */
export interface ScopeMarkerInfo {
  label: string;
  description: string;
  undo_description: string;
}

export interface MailActionInfo {
  /** Der Zustand, in dem die Zeile danach steht – die Aktion *ist* ihr Ziel. */
  id: MailState;
  label: string;
  description: string;
  tone: "positive" | "neutral" | "negative";
}

export interface MailEntry {
  contact_id: string;
  betrieb: string;
  telefon: string;
  email: string;
  ort: string;
  plz: string;
  website: string;
  gewerk: string;
  /** Die Prio aus der Datei – die Liste zeigt sie nicht, das Kopieren nimmt sie mit. */
  prio: string;
  /** Der Gesprächsaufhänger aus der Datei. */
  befunde: string;
  /**
   * Alle übrigen Spalten der Anrufliste, in der Reihenfolge der Datei.
   *
   * Angezeigt wird davon nichts – sie sind der Grund, warum eine Zeile sich
   * als Text kopieren lässt: was in ihnen steht, weiß nur die Analyse, aus
   * der die Liste kam.
   */
  extras: ContactField[];
  list_id: string;
  list_name: string;
  /** Archivierte Listen bleiben sichtbar – die Zusage gilt weiter. */
  list_archived: boolean;
  promised_at: string | null;
  promised_by: string;
  /** Anmerkung aus dem Telefonat. */
  note: string;
  state: MailState;
  state_label: string;
  /**
   * Die Bau-Einschätzung – unabhängig vom Versandstand, aber nur, solange
   * die Mail nicht heraus ist: danach `unbewertet`, weil die Reihe der Weg
   * *bis* zur Mail ist. Der gesetzte Wert bleibt gespeichert und kommt
   * zurück, wenn der Versand zurückgesetzt wird.
   */
  readiness: BuildReadiness;
  readiness_label: string;
  /**
   * Welche Marker diese Zeile setzen kann – leer, sobald die Mail heraus ist.
   *
   * Dieselbe Rolle wie `actions`: die Oberfläche baut die Regel nicht nach,
   * sonst wächst ihr ein Knopf, der mit 400 antwortet.
   */
  readiness_actions: BuildReadiness[];
  /**
   * „Bigger than expected" – die dritte Größe, kombinierbar mit den beiden
   * anderen. `false` heißt „hat niemand gesagt", nicht „ist ein Onepager".
   */
  oversized: boolean;
  /** Der Zustand folgt aus der Frist und wurde nicht angeklickt. */
  automatic: boolean;
  sent_at: string | null;
  answered_at: string | null;
  /** Wann telefonisch nachgefasst wurde – `null`, solange niemand angerufen hat. */
  followed_up_at: string | null;
  days_since_sent: number | null;
  /** Anmerkung zum Versand – getrennt von der aus dem Telefonat. */
  mail_note: string;
  updated_at: string | null;
  updated_by: string;
  /**
   * Welche Knöpfe diese Zeile zeigt.
   *
   * Kommt aus derselben Übergangstabelle, gegen die das Backend beim
   * Schreiben prüft. Die Oberfläche darf sie nicht selbst nachbauen, sonst
   * wächst ihr ein Knopf, der mit 400 antwortet.
   */
  actions: MailState[];
}

/**
 * Die Zahlen über der Liste – zwei Aufteilungen derselben Menge.
 *
 * Die Oberfläche hat zwei Filterreihen, und **jede zählt innerhalb der
 * Auswahl der anderen**: mit Reiter „Offen" nennen die Marker die offenen
 * Zusagen, und ihre Zahlen ergeben zusammen die Zahl auf dem Reiter.
 * Ihren eigenen Filter lässt eine Reihe außen vor, sonst gäbe es keinen
 * Rückweg, der eine Zahl nennt. Die Suche bleibt aus beiden heraus.
 *
 * Gerechnet wird das im Backend (`_counters`) – hier ist nur zu wissen, dass
 * die Zahlen zur mitgeschickten Sicht gehören.
 */
export interface MailCounters {
  /** Alle Zusagen der Auswahl – die Zahl hinter dem Reiter „Alle". */
  gesamt: number;
  offen: number;
  versendet: number;
  /** Fällig zum Anrufen – die zweite Zahl, die auf null laufen soll. */
  nachfassen: number;
  /** Angerufen, wartet weiter. */
  nachgefasst: number;
  positiv: number;
  abgelehnt: number;
  keine_antwort: number;
  /** Zusagen ohne Adresse – die Nacharbeit, die sonst niemand sieht. */
  ohne_email: number;
  /**
   * Die Bau-Einschätzung, gezählt innerhalb des Versandstand-Filters.
   *
   * Ohne ihn über alle Zusagen: „wie viele könnten wir sofort bauen?" ist die
   * Frage, für die die Marker gesetzt werden – und die stellt sich vor der
   * Antwort auf die Mail.
   */
  ready_to_build: number;
  /** Wird gerade gebaut. */
  in_development: number;
  /** Gebaut, aber noch nicht beim Betrieb. */
  ready_to_mail: number;
  missing_content: number;
  unbewertet: number;
  /** Zusagen, deren Seite größer als ein Onepager ist. Eine Zahl statt einer
   *  Aufteilung: der Marker ist gesetzt oder nicht. */
  oversized: number;
}

export interface MailBoard {
  /** Zählt jede Änderung an der Anrufdatenbank – steuert den Poll. */
  revision: number;
  counters: MailCounters;
  entries: MailEntry[];
  /** Zusagen insgesamt … */
  total: number;
  /** … und die, die Suche und Filter übrig lassen. */
  matched: number;
  offset: number;
  limit: number;
  actions: MailActionInfo[];
  /** Die Marker samt Beschriftung – wie `actions` Daten und nicht Code. */
  readiness_options: ReadinessOptionInfo[];
  /** Der Umfangs-Marker. Einer statt einer Liste: gesetzt oder nicht. */
  scope_marker: ScopeMarkerInfo;
  /** Die Frist, nach der ohne Antwort „keine Antwort" gilt. */
  timeout_days: number;
  /** Die kürzere Frist, nach der nachtelefoniert werden soll. */
  followup_days: number;
}

/** Die Sicht, aus der ein Klick kam – sie reist mit, damit die Liste steht. */
export interface MailView {
  q?: string;
  state?: MailState | null;
  /** Zweiter, unabhängiger Filter: „was ist verschickt" und „was können wir
   *  bauen" sind zwei Fragen, zusammen ergeben sie die Bauliste. */
  readiness?: BuildReadiness | null;
  /** Dritter Filter: „was ist größer als ein Onepager". `null` heißt „alle";
   *  `false` ist eine echte Auswahl (die Seiten ohne Marker). */
  oversized?: boolean | null;
  offset?: number;
  limit?: number;
}

/**
 * Was ein Klick schickt. Jedes Feld einzeln, jedes Weglassen „unverändert".
 *
 * Ohne `state` ist es eine reine Anmerkung oder ein reiner Marker – der
 * Versandstand bleibt dabei stehen, samt Versand- und Antwortdatum.
 */
export interface MailUpdate {
  state?: MailState;
  /** `undefined` = unverändert, `""` = löschen. */
  note?: string;
  /** `undefined` = unverändert; entfernt wird mit `"unbewertet"`. */
  readiness?: BuildReadiness;
  /** `undefined` = unverändert. Bei zwei Werten ist `false` der Rückweg – ein
   *  „unbewertet" wie bei der Bau-Einschätzung braucht es hier nicht. */
  oversized?: boolean;
}

/** Query-Parameter aus einer Sicht – leere Felder bleiben weg. */
function params(view: MailView): Record<string, string | number> {
  const query: Record<string, string | number> = {};

  if (view.q?.trim()) query.q = view.q.trim();
  if (view.state) query.state = view.state;
  if (view.readiness) query.readiness = view.readiness;
  // Nicht `if (view.oversized)`: `false` ist hier eine Auswahl und kein
  // „nicht gesetzt" – nur `null`/`undefined` heißt „alle".
  if (view.oversized !== null && view.oversized !== undefined) {
    query.oversized = String(view.oversized);
  }
  if (view.offset) query.offset = view.offset;
  if (view.limit) query.limit = view.limit;

  return query;
}

export async function fetchBoard(view: MailView = {}, signal?: AbortSignal): Promise<MailBoard> {
  const response = await http.get<MailBoard>("/mailversand/board", {
    params: params(view),
    signal,
  });
  return response.data;
}

/**
 * Setzt den Versandzustand einer Zusage – oder nur ihre Anmerkung.
 *
 * Antwortet mit der ganzen Ansicht für dieselbe Sicht, aus der der Klick kam;
 * deshalb reisen Suche, Filter und Seite als Query-Parameter mit. Ohne das
 * spränge die Liste nach jedem Klick auf die erste Seite zurück.
 */
export async function updateEntry(
  contactId: string,
  update: MailUpdate,
  view: MailView = {},
): Promise<MailBoard> {
  const response = await http.post<MailBoard>(`/mailversand/contacts/${contactId}`, update, {
    params: params(view),
  });
  return response.data;
}

/** Die ganze Versandliste als CSV. */
export function exportUrl(): string {
  return "/api/mailversand/export";
}

File: backend/app/api/mail_followup_api.py
"""HTTP-Schicht des Mailversands.

Fünf Endpunkte: die Ansicht lesen, einen Zustand setzen, einen Betrieb von
Hand anlegen, dessen Stammdaten ändern, alles als CSV herausholen. Der schreibende Aufruf antwortet mit der ganzen Ansicht, wie
überall in dieser Anwendung — und nimmt dafür Suche, Filter und Seite als
Query-Parameter entgegen, damit die Liste nach dem Klick dort stehen bleibt,
wo sie war (dasselbe Verfahren wie beim Freigeben einer gesperrten Nummer).

Der Zugang hängt an der eigenen Seitenberechtigung `mailversand` und
ausdrücklich **nicht** zusätzlich an `require_admin`: wer Mails versendet,
muss dafür kein Administrator sein — und wem die Seite nicht zugeteilt ist,
für den existiert sie nicht. Die Seitenberechtigungen sind das Mittel dieser
Anwendung für genau diese Frage; ein zweiter Riegel darüber machte sie zur
Attrappe.

Die Handler sind einfaches `def`: SQLite blockiert, und FastAPI führt
synchrone Handler in seinem Threadpool aus.
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse

from app.api.deps import CurrentUser, current_user
from app.schemas.access import Page
from app.schemas.mail_followup import (
    MAIL_PAGE_SIZE,
    MAX_MAIL_PAGE_SIZE,
    BuildReadiness,
    MailBoard,
    MailState,
    MailUpdateRequest,
    ManualEntryRequest,
)
from app.services.mail_followup_service import (
    MailFollowupConflictError,
    MailFollowupError,
    MailFollowupNotFoundError,
    create_entry,
    export_board,
    get_board,
    set_state,
    update_entry,
)

# Berechtigung, hinter der dieser Router hängt. `main.py` liest die Konstante
# beim Einhängen — ein Feature-Modul ohne sie lässt sich gar nicht erst
# mounten.
PAGE = Page.MAILVERSAND

router = APIRouter(prefix="/mailversand", tags=["mailversand"])


def _fail(exc: MailFollowupError) -> HTTPException:
    if isinstance(exc, MailFollowupNotFoundError):
        status = 404
    elif isinstance(exc, MailFollowupConflictError):
        status = 409
    else:
        status = 400

    return HTTPException(status_code=status, detail=str(exc))


@router.get("/board", response_model=MailBoard)
def read_board(
    q: str = Query(default="", max_length=200),
    state: MailState | None = Query(default=None),
    readiness: BuildReadiness | None = Query(default=None),
    oversized: bool | None = Query(default=None),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=MAIL_PAGE_SIZE, ge=1, le=MAX_MAIL_PAGE_SIZE),
    _: CurrentUser = Depends(current_user),
) -> MailBoard:
    """Die Versandliste: Zähler, eine Seite Zusagen, die Knöpfe.

    `state`, `readiness` und `oversized` sind drei unabhängige Filter, weil
    sie drei Fragen beantworten: was ist verschickt, was lässt sich bauen,
    und was ist größer als ein Onepager.
    """
    return get_board(
        query=q,
        state=state,
        readiness=readiness,
        oversized=oversized,
        offset=offset,
        limit=limit,
    )


@router.post("/contacts/{contact_id}", response_model=MailBoard)
def change_state(
    contact_id: str,
    request: MailUpdateRequest,
    q: str = Query(default="", max_length=200),
    state: MailState | None = Query(default=None),
    readiness: BuildReadiness | None = Query(default=None),
    oversized: bool | None = Query(default=None),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=MAIL_PAGE_SIZE, ge=1, le=MAX_MAIL_PAGE_SIZE),
    user: CurrentUser = Depends(current_user),
) -> MailBoard:
    """Einen Versandzustand, eine Anmerkung oder einen Marker setzen.

    Wer geklickt hat, kommt aus der Sitzung und nie aus dem Anfragekörper —
    dieselbe Regel wie im Anrufprotokoll, auch wenn hier kein Nachweis
    entsteht, sondern Arbeitsstand.

    `state`/`readiness`/`oversized` sind hier zweimal da und meinen
    zweierlei: in der Query die Sicht, die zurückkommen soll, im Körper das,
    was gesetzt wird.
    """
    try:
        return set_state(
            contact_id,
            request,
            username=user.username,
            query=q,
            state=state,
            readiness=readiness,
            oversized=oversized,
            offset=offset,
            limit=limit,
        )
    except MailFollowupError as exc:
        raise _fail(exc) from exc


@router.post("/contacts", response_model=MailBoard, status_code=201)
def create_contact(
    request: ManualEntryRequest,
    q: str = Query(default="", max_length=200),
    state: MailState | None = Query(default=None),
    readiness: BuildReadiness | None = Query(default=None),
    oversized: bool | None = Query(default=None),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=MAIL_PAGE_SIZE, ge=1, le=MAX_MAIL_PAGE_SIZE),
    user: CurrentUser = Depends(current_user),
) -> MailBoard:
    """Einen Betrieb von Hand anlegen — für die, die niemand angerufen hat.

    Antwortet mit der ganzen Ansicht, wie jeder Schreibzugriff hier. 409, wenn
    die Nummer schon bekannt ist: die Meldung nennt, wo sie steht, und
    `force` im Körper übergeht den Befund danach ausdrücklich.
    """
    try:
        return create_entry(
            request,
            user_id=user.id,
            username=user.username,
            query=q,
            state=state,
            readiness=readiness,
            oversized=oversized,
            offset=offset,
            limit=limit,
        )
    except MailFollowupError as exc:
        raise _fail(exc) from exc


@router.patch("/contacts/{contact_id}", response_model=MailBoard)
def edit_contact(
    contact_id: str,
    request: ManualEntryRequest,
    q: str = Query(default="", max_length=200),
    state: MailState | None = Query(default=None),
    readiness: BuildReadiness | None = Query(default=None),
    oversized: bool | None = Query(default=None),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=MAIL_PAGE_SIZE, ge=1, le=MAX_MAIL_PAGE_SIZE),
    user: CurrentUser = Depends(current_user),
) -> MailBoard:
    """Die Stammdaten eines von Hand erfassten Betriebs ändern.

    PATCH statt POST, weil unter derselben Adresse schon der Versandstand
    geschrieben wird: die Methode trennt „was ist aus der Zusage geworden" von
    „wer ist dieser Betrieb". Importierte Zeilen lehnt der Service ab — die
    gehören der Telefonakquise.
    """
    try:
        return update_entry(
            contact_id,
            request,
            user_id=user.id,
            username=user.username,
            query=q,
            state=state,
            readiness=readiness,
            oversized=oversized,
            offset=offset,
            limit=limit,
        )
    except MailFollowupError as exc:
        raise _fail(exc) from exc


@router.get("/export")
def download_board(_: CurrentUser = Depends(current_user)) -> StreamingResponse:
    """Die ganze Versandliste als CSV."""
    result = export_board()

    return StreamingResponse(
        result.buffer,
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="{result.filename}"'},
    )
